File: crates/breez-sdk/bindings/examples/cli/langs/dart/lib/contacts.dart
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';

import 'serialization.dart';

/// Contacts subcommand names (used for tab completion).
const contactsCommandNames = ['contacts add', 'contacts update', 'contacts delete', 'contacts list'];

typedef ContactsHandler = Future<void> Function(BreezSdk sdk, List<String> args);

class _ContactsEntry {
  final String description;
  final ContactsHandler handler;
  const _ContactsEntry(this.description, this.handler);
}

Map<String, _ContactsEntry>? _registry;

Map<String, _ContactsEntry> _getRegistry() {
  return _registry ??= {
    'add': _ContactsEntry('Add a new contact', _handleAdd),
    'update': _ContactsEntry('Update an existing contact', _handleUpdate),
    'delete': _ContactsEntry('Delete a contact', _handleDelete),
    'list': _ContactsEntry('List contacts', _handleList),
  };
}

/// Dispatch a contacts subcommand given the args after 'contacts'.
Future<void> dispatchContactsCommand(List<String> args, BreezSdk sdk) async {
  final registry = _getRegistry();

  if (args.isEmpty || args[0] == 'help' || args[0] == '--help') {
    print('\nContacts subcommands:\n');
    for (final entry in registry.entries.toList()..sort((a, b) => a.key.compareTo(b.key))) {
      print('  contacts ${entry.key.padRight(30)} ${entry.value.description}');
    }
    print('');
    return;
  }

  final subName = args[0];
  final subArgs = args.sublist(1);

  if (!registry.containsKey(subName)) {
    print("Unknown contacts subcommand: $subName. Use 'contacts help' for available commands.");
    return;
  }

  await registry[subName]!.handler(sdk, subArgs);
}

// --- add ---

Future<void> _handleAdd(BreezSdk sdk, List<String> args) async {
  if (args.length < 2 || args.first == 'help' || args.first == '--help') {
    print('Usage: contacts add <name> <payment_identifier>');
    return;
  }
  final name = args[0];
  final paymentIdentifier = args[1];
  final result = await sdk.addContact(
    request: AddContactRequest(name: name, paymentIdentifier: paymentIdentifier),
  );
  printValue(result);
}

// --- update ---

Future<void> _handleUpdate(BreezSdk sdk, List<String> args) async {
  if (args.length < 3 || args.first == 'help' || args.first == '--help') {
    print('Usage: contacts update <id> <name> <payment_identifier>');
    return;
  }
  final id = args[0];
  final name = args[1];
  final paymentIdentifier = args[2];
  final result = await sdk.updateContact(
    request: UpdateContactRequest(id: id, name: name, paymentIdentifier: paymentIdentifier),
  );
  printValue(result);
}

// --- delete ---

Future<void> _handleDelete(BreezSdk sdk, List<String> args) async {
  if (args.isEmpty || args.first == 'help' || args.first == '--help') {
    print('Usage: contacts delete <id>');
    return;
  }
  final id = args[0];
  await sdk.deleteContact(id: id);
  print('Contact deleted successfully');
}

// --- list ---

Future<void> _handleList(BreezSdk sdk, List<String> args) async {
  int? offset;
  int? limit;
  if (args.isNotEmpty && args.first != 'help' && args.first != '--help') {
    offset = int.tryParse(args[0]);
    if (args.length > 1) {
      limit = int.tryParse(args[1]);
    }
  } else if (args.isNotEmpty) {
    print('Usage: contacts list [offset] [limit]');
    return;
  }
  final result = await sdk.listContacts(request: ListContactsRequest(offset: offset, limit: limit));
  printValue(result);
}

File: crates/breez-sdk/bindings/examples/cli/langs/python/src/breez_cli/contacts.py
import argparse

from breez_sdk_spark import (
    AddContactRequest,
    ListContactsRequest,
    UpdateContactRequest,
)

from breez_cli.serialization import print_value

# Contacts subcommand names (used for REPL completion)
CONTACTS_COMMAND_NAMES = [
    "contacts add",
    "contacts update",
    "contacts delete",
    "contacts list",
]


def _parser(name, description=""):
    return argparse.ArgumentParser(prog=f"contacts {name}", description=description)


# --- add ---

def _build_add_parser():
    p = _parser("add", "Add a new contact")
    p.add_argument("name", help="Name of the contact")
    p.add_argument("payment_identifier", help="Lightning address (user@domain)")
    return p

async def _handle_add(sdk, args):
    result = await sdk.add_contact(
        request=AddContactRequest(
            name=args.name,
            payment_identifier=args.payment_identifier,
        )
    )
    print_value(result)


# --- update ---

def _build_update_parser():
    p = _parser("update", "Update an existing contact")
    p.add_argument("id", help="ID of the contact to update")
    p.add_argument("name", help="New name for the contact")
    p.add_argument("payment_identifier", help="New Lightning address (user@domain)")
    return p

async def _handle_update(sdk, args):
    result = await sdk.update_contact(
        request=UpdateContactRequest(
            id=args.id,
            name=args.name,
            payment_identifier=args.payment_identifier,
        )
    )
    print_value(result)


# --- delete ---

def _build_delete_parser():
    p = _parser("delete", "Delete a contact")
    p.add_argument("id", help="ID of the contact to delete")
    return p

async def _handle_delete(sdk, args):
    await sdk.delete_contact(id=args.id)
    print("Contact deleted successfully")


# --- list ---

def _build_list_parser():
    p = _parser("list", "List contacts")
    p.add_argument("offset", nargs="?", type=int, default=None,
                   help="Number of contacts to skip")
    p.add_argument("limit", nargs="?", type=int, default=None,
                   help="Maximum number of contacts to return")
    return p

async def _handle_list(sdk, args):
    result = await sdk.list_contacts(
        request=ListContactsRequest(
            offset=args.offset,
            limit=args.limit,
        )
    )
    print_value(result)


# ---------------------------------------------------------------------------
# Registry and dispatch
# ---------------------------------------------------------------------------

def _build_contacts_registry():
    return {
        "add": (_build_add_parser(), _handle_add),
        "update": (_build_update_parser(), _handle_update),
        "delete": (_build_delete_parser(), _handle_delete),
        "list": (_build_list_parser(), _handle_list),
    }


_REGISTRY = None

def _get_registry():
    global _REGISTRY
    if _REGISTRY is None:
        _REGISTRY = _build_contacts_registry()
    return _REGISTRY


async def dispatch_contacts_command(args, sdk):
    """Dispatch a contacts subcommand given the args after 'contacts'."""
    registry = _get_registry()

    if not args or args[0] == "help":
        print("\nContacts subcommands:\n")
        for name, (parser, _) in sorted(registry.items()):
            desc = parser.description or ""
            print(f"  contacts {name:30s} {desc}")
        print()
        return

    sub_name = args[0]
    sub_args = args[1:]

    if sub_name not in registry:
        print(f"Unknown contacts subcommand: {sub_name}. Use 'contacts help' for available commands.")
        return

    parser, handler = registry[sub_name]
    try:
        parsed = parser.parse_args(sub_args)
    except SystemExit:
        return

    await handler(sdk, parsed)
